content/recipes/async-local-storage.md

```diff
@@ -0,0 +1,227 @@
+### Async Local Storage
+
+`AsyncLocalStorage` is a [Node.js API](https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage) (based on the `async_hooks` API) that provides an alternative way of propagating local state through the application without the need to explicitly pass it as a function parameter. It is similar to a thread-local storage in other languages.
+
+The main idea of Async Local Storage is that we can _wrap_ some function call with the `AsyncLocalStorage#run` call. All code that is invoked within the wrapped call gets access to the same `store`, which will be unique to each call chain.
+
+In the context of NestJS, that means if we can find a place within the request's lifecycle where we can wrap the rest of the request's code, we will be able to access and modify state visible only to that request, effectively eliminating the need for REQUEST scoped providers and their limitations.
+
+#### Custom implementation
+
+NestJS itself does not provide any built-in abstraction for `AsyncLocalStorage`, so let's walk through how we could implement it ourselves for the simplest HTTP case to get a better understanding of the whole concept:
+
+> info **info** For a ready-made [dedicated package](recipes/async-local-storage#nestjs-cls), continue reading below.
+
+1. First, create a new instance of the `AsyncLocalStorage` in some shared source file. Since we're using NestJS, let's also turn it into a module with a custom provider.
+
+```ts
+/** als.module.ts */
+
+import { AsyncLocalStorage } from 'async_hooks';
+import { Module } from '@nestjs/core';
+
+export const asyncLocalStorage = new AsyncLocalStorage();
+
+@Module({
+  providers: [
+    {
+      provide: AsyncLocalStorage,
+      useValue: asyncLocalStorage,
+    },
+  ],
+  exports: [AsyncLocalStorage],
+})
+export class AlsModule {}
+```
+
+2. We're only concerned with HTTP, so let's use a middleware to wrap the `next` function with `AsyncLocalStorage#run`. Since a middleware is the first thing that the request hits, this will make the `store` available in all enhancers and the rest of the system.
+
+```ts
+/** main.ts */
+
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+import { asyncLocalStorage } from './als.setup.ts';
```

Comment:
why we have to import the instance here?

also, 
```suggestion
import { asyncLocalStorage } from './als.module.ts';
```

Reply:
Oh, I overlooked that when re-facring the example :)